### PR TITLE
Drop old semaphore implementation

### PR DESCRIPTION
**Pull request by RandomShaper**
- Removed platform-specific implementations.
- Now all semaphores are in-object, unless they need to be conditionally created.
- Similarly to `Mutex`, provided a dummy implementation for when `NO_THREADS` is defined.
- Similarly to `Mutex`, methods are made `const` for easy use in such contexts.
- Language bindings updated: `wait()` and `post()` are now `void`.
- Language bindings updated: `try_wait()` added.

### CODE REVIEW

File: core/os/semaphore.h [hunk 1]
```diff
@@ -34,30 +34,32 @@
 #include "core/error_list.h"
 #include "core/typedefs.h"
 
+#if !(defined NO_THREADS)
```
Comment (akien-mga) [inline core/os/semaphore.h]: We'd usually write this `#if !defined(NO_THREADS)`.

**Review by akien-mga — APPROVED** [review]
Looks good.

Might conflict with @reduz's OS/DisplayServer split, but conflicts shouldn't be hard to solve.

File: core/os/semaphore.h (hunk 1 above)
Comment (RandomShaper) [reply core/os/semaphore.h]: Fixed.

File: servers/physics_2d/physics_2d_server_wrap_mt.cpp [hunk 2]
```diff
@@ -84,11 +84,11 @@ void Physics2DServerWrapMT::step(real_t p_step) {
 
 void Physics2DServerWrapMT::sync() {
 
-	if (step_sem) {
+	if (thread) {
```
Comment (RandomShaper) [inline servers/physics_2d/physics_2d_server_wrap_mt.cpp]: @reduz, please validate this change.

For the semaphore to be "in-object", now it's created unconditionally, so this check couldn't be based in `step_sem` anymore.

The new logic is equivalent, isn't it?

Comment (reduz) [reply servers/physics_2d/physics_2d_server_wrap_mt.cpp]: its fine

**Comment by akien-mga** [comment]
Thanks!

File: servers/physics_2d/physics_2d_server_wrap_mt.cpp (hunk 2 above)
Comment (akien-mga) [reply servers/physics_2d/physics_2d_server_wrap_mt.cpp]: Comment from @reduz:
```
i think its ok, the wrapmt classes should probably be rewritten using spinlocks mainly
```

**Pull request merged**